tests/cases/compiler/deferredCallbacks.ts

```diff
@@ -0,0 +1,238 @@
+declare function f23<T extends string[] | (() => void)[]>(deferred ...funcs: T): void;
+declare function f24<T extends (() => void)[]>(deferred ...funcs: T | string[]): void;
+
+declare function f30(deferred f: { foo(): void }): void;
+declare function f31(deferred f: number): void;
+declare function f32(deferred ...funcs: number[]): void;
+
+type T10 = (deferred f: () => void) => void;
+type T11 = (deferred f: { (): void }) => void;
+type T12 = (deferred f: Function) => void;
+type T13 = (deferred f: any) => void;
+
+type T20 = (deferred f: { foo(): void }) => void;
+type T21 = (deferred f: number) => void;
+type T22 = (deferred ...funcs: number[]) => void;
+type T23 = { deferred x: () => void };
+
+// deferred modifier is not captured in argument list tuples
+
+declare function doStuff(deferred f: () => void): void;
+
+declare function recreate<A extends unknown[], R>(f: (...args: A) => R): (...args: A) => R;
+declare function recreateDeferred1<A extends unknown[], R>(f: (deferred ...args: A) => R): (...args: A) => R;
+declare function recreateDeferred2<A extends unknown[], R>(f: (...args: A) => R): (deferred ...args: A) => R;
+
+function ff1() {
+    let x: string | number;
+    x = 123;
+    doStuff(() => {
+        x = "hi";
+    });
+    x;  // number
+}
+
+function ff2() {
+    let y: string | number;
+    y = 123;
+    recreate(doStuff)(() => {
+        y = "hi";
+    });
+    y;  // string | number
+}
+
+function ff3() {
+    let z: string | number;
+    z = 123;
+    recreateDeferred1(doStuff)(() => {
+        z = "hi";
+    });
+    z;  // string | number
+}
+
+function ff4() {
+    let z: string | number;
+    z = 123;
+    recreateDeferred2(doStuff)(() => {
+        z = "hi";
+    });
+    z;  // number
+}
+
+// https://github.com/microsoft/TypeScript/issues/11498
+
+declare function mystery(cb: () => void): void;
+
+function fx1() {
+    let x: string | number = "OK";
+    x;  // string
+    mystery(() => {
+        x = 10;
+    });
+    x;  // string | number
+    if (x === 10) {}
+}
+
+// https://github.com/microsoft/TypeScript/issues/15380
+
+class Foo {
+  public bar: string = "";
+}
+
+function fx2() {
+  let foo: Foo | null = null;
+  [1].forEach((item) => {
+      foo = new Foo();
+  });
+  if (foo) {
+      foo.bar;
+  }
+}
+
+// https://github.com/microsoft/TypeScript/issues/57880
+
+const call = (f: () => void) => f();
+
+const fx3 = () => {
+    let a: undefined | number = undefined;
+    call(() => { a = 1; });
+    if (a !== undefined) {
+        a.toString();
+    }
+};
+
+// https://github.com/microsoft/TypeScript/issues/58291
+
+async function execute(onError: (_err: Error | undefined) => void) {
+    onError(new Error("a"));
+}
+
+async function run() {
+    let result: boolean = true;
+    await execute(() => {
+        result = false;
+    });
+    if (result === false) {
+        console.log("error");
+    }
+    return result;
+}

```

Comment:
Can you add a test like

```ts
class C {
  constructor(options?: {
    /** @deferred */ fn?: Function | undefined
  }) { }
}
```

I failed to get an error in this case in [my node PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70084)